internal/ppa/ppa.go

```diff
@@ -0,0 +1,377 @@
+
+type PPAInterface interface {
+	Add(basePath string, debug bool) error
+	Remove(string) error
+}
+
+// PPAPrivateInterface defines internal behavior expected from a PPAInterface implementer.
+// Even though methods are exporter, they are not meant to be used outside of this file.
+type PPAPrivateInterface interface {
+	FullName() string
+	FileName() string
+	FileContent() (string, error)
+	ImportKey(basePath string, debug bool) error
+	Remove(string) error
+}
+
+// New instantiate the proper PPA implementation based on the deb822 flag
+func New(imageDefPPA *imagedefinition.PPA, deb822 bool, series string) PPAInterface {
+	basePPA := BasePPA{
+		PPA:    imageDefPPA,
+		Series: series,
+	}
+
+	if deb822 {
+		return &PPA{
+			PPAPrivateInterface: &Deb822PPA{
+				BasePPA: basePPA,
+			},
+		}
+	}
+
+	return &PPA{
+		PPAPrivateInterface: &LegacyPPA{
+			BasePPA: basePPA,
+		},
+	}
+}
+
+// BasePPA holds fields and methods common to every PPAPrivateInterface implementation
+type BasePPA struct {
+	*imagedefinition.PPA
+	Series     string
+	signingKey string
+}
+
+func (p *BasePPA) FullName() string {
+	return p.Name
+}
+
+func (p *BasePPA) name() string {
+	return strings.Split(p.Name, "/")[1]
+}
+
+func (p *BasePPA) user() string {
+	return strings.Split(p.Name, "/")[0]
+}
+
+func (p *BasePPA) url() string {
+	var baseURL string
+	if p.Auth == "" {
+		baseURL = "https://ppa.launchpadcontent.net"
+	} else {
+		baseURL = fmt.Sprintf("https://%s@private-ppa.launchpadcontent.net", p.Auth)
+	}
+	return fmt.Sprintf("%s/%s/%s/ubuntu", baseURL, p.user(), p.name())
+}
+
+// removePPAFile removes the PPA file from the sources.list.d directory
+func (p *BasePPA) removePPAFile(basePath string, fileName string) error {
+	sourcesListD := filepath.Join(basePath, sourcesListDPath)
+	if p.KeepEnabled == nil {
+		return imagedefinition.ErrKeepEnabledNil
+	}
+
+	if *p.KeepEnabled {
+		return nil
+	}
+
+	ppaFile := filepath.Join(sourcesListD, fileName)
+	err := osRemove(ppaFile)
+	if err != nil {
+		return fmt.Errorf("Error removing %s: %s", ppaFile, err.Error())
+	}
+	return nil
+}
+
+func (p *BasePPA) importKey(basePath string, ppaFileName string, ascii bool, debug bool) (err error) {
+	trustedGPGD := filepath.Join(basePath, trustedGPGDPath)
+	keyFileName := strings.Replace(ppaFileName, ".list", ".gpg", 1)
+	keyFilePath := filepath.Join(trustedGPGD, keyFileName)
+
+	err = p.ensureFingerprint(lpBaseURL)
+	if err != nil {
+		return err
+	}
+
+	tmpGPGDir, err := p.createTmpGPGDir(basePath)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
+		tmpErr := osRemoveAll(tmpGPGDir)
+		if tmpErr != nil {
+			if err != nil {
+				err = fmt.Errorf("%s after previous error: %w", tmpErr.Error(), err)
+			} else {
+				err = fmt.Errorf("Error removing temporary gpg directory \"%s\": %s", tmpGPGDir, tmpErr.Error())
+			}
+		}
+	}()
+
+	tmpASCIIKeyFilName := keyFileName + ".asc"
+	tmpASCIIKeyFilePath := filepath.Join(tmpGPGDir, tmpASCIIKeyFilName)
+
+	commonGPGArgs := []string{
```

Comment:
Let's avoid the dance with fingerprint, gpg, and keyserver, and just fetch the key using the getSigningKeyData end point? As I mentioned earlier on Mattermost, URL looks like

`https://api.launchpad.net/devel/~juliank/+archive/ubuntu/apt-esm?ws.op=getSigningKeyData`

This returns a JSON string which contains the ASCII armored key